internal/controllers/v1beta1/gateway_test.go

```diff
@@ -2,20 +2,435 @@ package v1beta1
+			Kind:       "Certificate",
+			APIVersion: "cert-manager.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      TestCertificateName,
+			Namespace: TestCertNamespace,
+		},
+		Spec: v1certmanager.CertificateSpec{
+			DNSNames: []string{"test1.example.com", "test2.example.com"},
+			IssuerRef: v1.ObjectReference{
+				Kind:  "ClusterIssuer",
+				Name:  "default",
+				Group: "cert-manager.io",
+			},
+		},
+	}))
+
+	helper := NewTestHelperWithGateways(opts...)
+	return helper
+}
+
+func setupControllerWithSpy(cs *istiofake.Clientset, certFake *certmanagerfake.Clientset, opts *GatewayOptions) *controllerSpy {
+	spy := &controllerSpy{
+		GatewayController: NewGatewayController(cs, certFake, WithDryRun(opts.DryRun), WithCertificateNamespace(TestCertNamespace)),
+	}
+	spy.GatewayController.certHandler = spy
+	return spy
+}
+
+func (r *controllerSpy) CreateCertificate(ctx context.Context, gateway *v1beta1.Gateway, server *networkingv1beta1.Server) error {
+	r.CreateCalled++
+	if r.Error {
+		return fmt.Errorf("mock create error")
+	}
+	return r.GatewayController.CreateCertificate(ctx, gateway, server)
+}
+
+func (r *controllerSpy) UpdateCertificate(ctx context.Context, cert *v1certmanager.Certificate, gateway *v1beta1.Gateway, server *networkingv1beta1.Server) error {
+	r.UpdateCalled++
+	if r.Error {
+		return fmt.Errorf("mock update error")
+	}
+	return r.GatewayController.UpdateCertificate(ctx, cert, gateway, server)
+}
+
+func reconcileRequest() reconcile.Request {
+	return reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: TestNamespace,
+			Name:      TestGatewayName,
+		},
+	}
+}
+
+func assertCreateCertificateCalled(t *testing.T, helper *TestHelper) {
+	r, err := helper.Controller.Reconcile(context.TODO(), reconcileRequest())
+	assert.NoError(t, err)
+	assert.Equal(t, 1, helper.Controller.CreateCalled)
+	assert.Equal(t, reconcile.Result{}, r)
+}
+
+func assertCertificateUpdated(t *testing.T, helper *TestHelper) {
+	_, err := helper.Controller.Reconcile(context.TODO(), reconcileRequest())
+	assert.NoError(t, err)
+	assert.Equal(t, 0, helper.Controller.CreateCalled)
+	assert.Equal(t, 1, helper.Controller.UpdateCalled)
+}
+
+func TestGatewayControllerReconcileNoError(t *testing.T) {
+	t.Parallel()
+	g := NewGatewayController(istiofake.NewSimpleClientset(), certmanagerfake.NewSimpleClientset(), WithCertificateNamespace("default"))
 	r, err := g.Reconcile(context.TODO(), reconcile.Request{})
 	assert.NoError(t, err)
 	assert.Equal(t, reconcile.Result{}, r)
 }
+
+func TestGatewayReconcile_OkOnNotExists(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelper(NewGatewayOptions())
+	r, err := helper.Controller.Reconcile(context.TODO(), reconcileRequest())
+	assert.NoError(t, err)
+	assert.Equal(t, reconcile.Result{}, r)
+}
+
+func TestGatewayReconcile_CallsCreateCertificate(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelperWithGateways()
+	assertCreateCertificateCalled(t, helper)
+}
```

Comment:
This seems like the same test as `TestGatewyReconcile_CreatesCertificateForGateway`

Reply:
I'll rename this test. Its focused on showing a no-op given a nil Gateway.